### DISCUSSION TITLE
Fix icon fetching for some sites

### DISCUSSION THREAD

**Discussion by quexten** [dani-garcia/vaultwarden#3773] [Ideas]
A few sites in my vault (f.e rakuten.de, kraken.com, crunchyroll.com, etc.) do not have an icon in the clients. The current vaultwarden logic just checks a few well known paths (/favicon.ico etc.). These sites do not use these paths, but have the icon linked in the html:
`<link rel="apple-touch-icon" sizes="144x144" href="https://www.crunchyroll.com/build/assets/img/favicons/apple-touch-icon-144x144.png">`

We could parse the index.html and look for the icon link in order to fetch the items correctly. (I'm not sure if we have to take special protections to prevent ddos from large html sites. Maybe this should be an opt-in feature).

**Comment by quexten** [comment]
Actually I just noticed my brief skimming of the code was not very thorough and vaultwarden already parses the html. However, the sites mentioned above are still not working either way, so this might be a bug.

**Comment by stefan0xC** [comment]
@BlackDex recently improved the favicon handling https://github.com/dani-garcia/vaultwarden/pull/3751 so it might already work. You can test it with the latest `testing` image.

Also make sure you have not disabled the icon downloading (e.g. by setting `DISABLE_ICON_DOWNLOAD=true` or having it disabled in your web-vaults preferences).

**Reply by quexten** [reply]
Yeah, I'm testing on the latest development build and cleared out the icon cache to make sure. I've investigated the sites a bit;

- kraken.com seems to block requests without a known user-agent header
- rakuten.de does not have the favicons in the html, but in a manifest.json (https://web.dev/add-manifest/) - this should be fairly simple to implement and add favicon supports to some more modern sites
- for crunchyroll.com vaultwarden tries to get `/apple-touch-icon.png` but gets a 403 forbidden. I'm not sure why the icon fetching logic chose apple-touch-icon.png here, since the html does not contain this, only specific ones for various sizes: `build/assets/img/favicons/apple-touch-icon-114x114.png`.

So at least the handling of crunchyroll.com could be a bug. Support for manifest.json could be a feature.

**Reply by stefan0xC** [reply]
Just tested it with the current `main` branch and that is the only one that worked for me:

```
[2023-08-13 23:07:03.246][vaultwarden::api::icons][INFO] Downloaded icon from https://www.crunchyroll.com/build/assets/img/favicons/favicon-32x32.png
[2023-08-13 23:07:03.246][response][INFO] (icon_internal) GET /icons/<domain>/icon.png => 200 OK
```

**Reply by quexten** [reply]
```
[2023-08-13 21:19:05.599][reqwest::async_impl::client][DEBUG] redirecting 'http://crunchyroll.com/apple-touch-icon.png' to 'https://www.crunchyroll.com/apple-touch-icon.png'
[2023-08-13 21:19:05.599][h2::codec::framed_write][DEBUG] send frame=Headers { stream_id: StreamId(13), flags: (0x5: END_HEADERS | END_STREAM) }
[2023-08-13 21:19:05.615][h2::codec::framed_read][DEBUG] received frame=Headers { stream_id: StreamId(13), flags: (0x4: END_HEADERS) }
[2023-08-13 21:19:05.615][h2::codec::framed_read][DEBUG] received frame=Data { stream_id: StreamId(13) }
[2023-08-13 21:19:05.615][h2::codec::framed_read][DEBUG] received frame=Data { stream_id: StreamId(13), flags: (0x1: END_STREAM) }
[2023-08-13 21:19:05.615][vaultwarden::api::icons][DEBUG] Req.
[CAUSE] reqwest::Error {
    kind: Status(
        403,
    ),
    url: Url {
        scheme: "https",
        cannot_be_a_base: false,
        username: "",
        password: None,
        host: Some(
            Domain(
                "www.crunchyroll.com",
            ),
        ),
        port: None,
        path: "/apple-touch-icon.png",
        query: None,
        fragment: None,
    },
}
[2023-08-13 21:19:05.615][vaultwarden::api::icons][WARN] Unable to download icon: Empty response or unable find a valid icon. crunchyroll.com
```

### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.29.1-3d2df6ce
* Web-vault version: v2023.7.1
* OS/Arch: linux/aarch64
* Running within Docker: true (Base: Alpine)
* Environment settings overridden: false
* Uses a reverse proxy: true
* IP Header check: true (X-Forwarded-For)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Browser/Server Time Check: true
* Server/NTP Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.41.2
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** 


```json
{
  "_duo_akey": null,
  "_enable_duo": false,
  "_enable_email_2fa": true,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_icon_service_csp": "",
  "_icon_service_url": "",
  "_ip_header_enabled": true,
  "_smtp_img_src": "cid:",
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_session_lifetime": 20,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "auth_request_purge_schedule": "30 * * * * *",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_conn_init": "",
  "database_max_conns": 10,
  "database_timeout": 30,
  "database_url": "***************",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://*********************",
  "domain_origin": "*****://*********************",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": true,
  "emergency_notification_reminder_schedule": "0 3 * * * *",
  "emergency_request_timeout_schedule": "0 7 * * * *",
  "enable_db_wal": true,
  "event_cleanup_schedule": "0 10 0 * * *",
  "events_days_retain": null,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 60,
  "icon_cache_ttl": 86400,
  "icon_download_timeout": 60,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "invitation_expiration_hours": 120,
  "invitation_org_name": "Quexlabs",
  "invitations_allowed": true,
  "ip_header": "X-Forwarded-For",
  "job_poll_interval_ms": 30000,
  "log_file": "/data/vaultwarden.log",
  "log_level": "DEBUG",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": null,
  "org_creation_users": "",
  "org_events_enabled": false,
  "org_groups_enabled": false,
  "password_hints_allowed": true,
  "password_iterations": 100000,
  "push_enabled": true,
  "push_installation_id": "***",
  "push_installation_key": "***",
  "push_relay_uri": "https://push.bitwarden.com",
  "reload_templates": false,
  "require_device_email": true,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sendmail_command": null,
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": true,
  "signups_allowed": true,
  "signups_domains_whitelist": "",
  "signups_verify": true,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": false,
  "smtp_embed_images": true,
  "smtp_explicit_tls": null,
  "smtp_from": "*********************",
  "smtp_from_name": "Bitwarden",
  "smtp_host": "*******************",
  "smtp_password": "***",
  "smtp_port": 587,
  "smtp_security": "starttls",
  "smtp_ssl": null,
  "smtp_timeout": 15,
  "smtp_username": "***********",
  "templates_folder": "data/templates",
  "tmp_folder": "data/tmp",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_sendmail": false,
  "use_syslog": false,
  "user_attachment_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": false,
  "websocket_port": 3012,
  "yubico_client_id": "54614",
  "yubico_secret_key": "***",
  "yubico_server": null
}
```
</details>

**Reply by quexten** [reply]
Running the latest `testing-alpine`.

**Reply by quexten** [reply]
A select few more that do not work:
- camelcamelcamel.com
- codepen.io
- coolblue.de
- my.konami.net
- rebuy.de

**Reply by BlackDex** [reply]
Thanks for the sites. I will take a look at them when i have time.
Some sites have strange html layouts.

I have a bunch of test sites i use and also compare it with Bitwarden, Google and DDG.

I quickly checked kraken.com and crunchyroll.com, they don't seem to work on the Bitwarden, Google and other icon fetching tool either.

Also, some sites have Cloudflare protection enabled, and those block access since they require JavaScript.

The reason you see it tries to fetch `/apple-touch-icon.png` is because that, together with `/favicon.ico` are fallback's which will be tried if nothing was able to be extracted from the parsed HTML.

Unfortunately for some sites we are just not able to fetch icons from.
Ill try to see if i can dive a bit deeper into the mentioned sites, but it might just be not possible.

Trying to use `manifest.json` adds more checks and more requests coming from the server. It might be an option to add this, i did had this in mind already, but it will add extra burden and complexity, and it needs more testing then the changes i made just a few weeks ago.

Also, we can't really change the user-agent. The one currently used helps bypassing some sites via Cloudflare with some protection enabled. Adding a different try with a different UA would also make it take downloading icons very long and complex in my opinion. But maybe it's something to look at.

So, it might be some sites just will not work because of protections or other strange way of html or JavaScript stuff.

**Reply by quexten** [reply]
> Also, some sites have Cloudflare protection enabled, and those block access since they require JavaScript.

Just noticed this too while debugging. When accessing the site from the server I host bitwarden on, I get a captcha.

> Trying to use manifest.json adds more checks and more requests coming from the server. It might be an option to add this, i did had this in mind already, but it will add extra burden and complexity, and it needs more testing then the changes i made just a few weeks ago.

Yeah... rakuten.de specifically also offers svg directly but this would also require some svg2png library/renderer and probably be too complex. The html does link to the manifest, so at least we would only fetch it if the site actually has a manifest.

> Also, we can't really change the user-agent. The one currently used helps bypassing some sites via Cloudflare with some protection enabled. Adding a different try with a different UA would also make it take downloading icons very long and complex in my opinion. But maybe it's something to look at.

Agreed.

**Reply by BlackDex** [reply]
> The html does link to the manifest, so at least we would only fetch it if the site actually has a manifest.

That still adds more complexity though. I have the same issue with sites that use `http-equiv` to redirect.

We would need to parse it, see if there are no link items at all, if so redirect/fetch again, keep track of the redirect amounts etc...
It is something on my to-do list though. But it is not something which really breaks the UX i think. It will enhance it though.

**Reply by BlackDex** [reply]
I did some quick testing:

- kraken.com > Needs a security check/Javascript
- rakuten.de > Provides a SVG only, which i deem unsafe to be implemented directly, and converting to a PNG is in my eyes a bit too much.
- www.crunchyroll.com > Works
- camelcamelcamel.com > Works
- codepen.io > Works
- coolblue.de > Cloudflare Block/Javascript
- my.konami.net > Invalid certificate
- rebuy.de > Works


A workaround would be to manually add an icon to the icon cache using the same naming convention as you see in the cache.
Other than that, it would be hard to add these.
The only item which might be interesting would be the manifest part. But, the site you mentioned which has that actually just needs to fix there HTML, since they provide a `apple-touch-icon` `rel`, but that points to an invalid, not existing image.
